includes/discovery/sensors/state/himoinsa-gensets.inc.php

```diff
@@ -0,0 +1,368 @@
+<?php
+/**
+ * himoinsa-gensets.inc.php
+ *
+ * LibreNMS state sensor discovery module for Himoinsa Gensets
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * @link       https://www.librenms.org
+ *
+ * @copyright  2021 Daniel Baeza
+ * @author     TheGreatDoc <doctoruve@gmail.com>
+ */
+
+/*
+CEC7 (a.k.a CEA7CC2) bitmap status ( R G B T Mn Au Al)
+statusConm.0
+R = Mains commutator closed
+G = Gen commutator closed
+B = Blocked mode
+T = Test mode
+Mn = Manual mode
+Au = Auto mode
+Al = Active commutator alarm
+
+In LibreNMS it equals to 3 state sensors:
+Closed commutator: Mains or Genset (what commutator is closed, so where the power comes from)
+Genset Mode: Block, Test, Manual, Auto (the 4 modes of the genset)
+Alarm: Yes or No (If there is a commutator alarm)
+
+Example:
+Value = 66
+Value binary = 1000010
+States equals to:
+- Alarm: No active alarm
+- Genset Mode: Auto
+- Closed commutator: Mains
+
+CEA7 / CEM7 (CEA7 is a combination of CEC7 + CEM7 in a single Central) bitmap status (R G Al Bt B T Mn Au P A)
+status.0
+R = Mains commutator closed
+G = Gen commutator closed
+Al = Active Alarm
+Bt = Transfer Pump
+B = Blocked mode
+T = Test mode
+Mn = Manual mode
+Au = Auto mode
+P = Motor Stopped
+A = Motor Running
+*/
+$status = SnmpQuery::get(['HIMOINSAv14-MIB::status.0', 'HIMOINSAv14-MIB::statusConm.0'])->values();
+var_dump($status);
+// map of bits
+if (isset($status['HIMOINSAv14-MIB::status.0'])) {
+    $map = [
+        'Mains commutator closed' => 512,
+        'Gens commutator closed' => 256,
+        'Active Alarm' => 128,
+        'Transfer Pump' => 64,
+        'Blocked mode' => 32,
+        'Test mode' => 16,
+        'Manual mode' => 8,
+        'Auto mode' => 4,
+        'Motor Stopped' => 2,
+        'Motor Running' => 1,
+    ];
+
+    $statusMotor = 0;
+    $statusMode = 0;
+    $statusAlarm = 0;
+    $statusComm = 0;
+    $statusTransferPump = 0;
+    $statusCommAlarm = 0;
+
+    foreach ($map as $descr => $mask) {
```

Comment:
Still confusing, why are we looping, why don't we just check each mask once and set the variable?

Reply:
Like this?

```php
$status = $status['HIMOINSAv14-MIB::status.0'] ?? 0;
$statusMotor = $status & 512 | $status & 256;  // assuming these values are mutually exclisive, it will result in a value of 256, 512, or 0;  (768 if both can be set)
```

```

Reply:
Its a feature. If not, the code is going to be so fast and clean that some servers could implode and generate a micro black hole, destroying the whole solar system.